TokensController: Memoize strategy.authorize_response result

pjambet opened the pull request:
This allows easy access to the authorize response in a subclass of Doorkeeper::TokensController.
Without this change, I had to user the following code to access the authorize response:

```ruby
strategy.request.instance_variable_get('@response')
```

If you accept this PR, I'll update the commit to include the CHANGELOG entry, I simply didn't want to add that if you think this PR should never get merged.
I'm not 100% satisfied with the spec I added, but at least it gets the job done (It's red without the change, and green with the change), let me know if you want to improve it.

Another alternative I had in mind was to return `response` from `TokensController#create`, but I thought memoizing the result was a better solution.

Let me know if you have any questions.

tute commented on app/controllers/doorkeeper/tokens_controller.rb:
```diff
@@ -37,5 +37,9 @@ def revoke_token(token)
     end
+
+    def authorize_response
+      @authorize_response ||= strategy.authorize
```
We don't need the `@authorize_response ||= ` memoization part.

pjambet replied:
Actually this is the main reason I sent this pull request, because `strategy.authorize` is not idempotent, if in a subclass of `TokensController` I need the response, I cannot call `strategy.authorize` again, therefore the need for the memoization.

tute replied:
Got it. Please explain this in the commit message too. :)

tute replied:
And when we have the CHANGELOG updated and a good commit message, I'll merge.

pjambet replied:
Awesome, going to update this to make it more explicit, and add the CHANGELOG entry, that should be ready by tomorrow.

Thanks!

tute commented on spec/controllers/tokens_controller_spec.rb:
```diff
@@ -40,4 +40,18 @@
+      end
+      post :create
+    end
+  end
```
And if we don't memoize, we don't need this spec.

tute commented:
I'm happy with this extract method, if it makes working with doorkeeper easier for you. Your child controller could get the response calling the method: `authorize_response`. I'm fine with it not being memoized.

Please explain in the commit message and CHANGELOG that this allows for subclasses to get the response object to do whatever they need, with a concrete example if you have, and I'll merge.

Thanks!

pjambet commented:
I updated the commit message and added a CHANGELOG entry. I tried to keep the CHANGELOG entry pretty short as the commit contains all the details.

Let me know if you want to change anything before merging this.

tute commented:
That's great, thank you!

pjambet commented:
Thanks for merging this! And kudos for the gem, it's been a pleasure to work with so far.
